src/_pytest/main.py

```diff
@@ -491,14 +493,26 @@ def gethookproxy(self, fspath):
             proxy = self.config.hook
         return proxy
 
+    def pytest_deselected(self, items):
+        """Keep track of explicitly deselected items."""
+        self._deselected.extend(items)
+
     def perform_collect(self, args=None, genitems=True):
         hook = self.config.hook
         try:
             items = self._perform_collect(args, genitems)
             self.config.pluginmanager.check_pending()
+
+            items_before = set(items)
             hook.pytest_collection_modifyitems(
                 session=self, config=self.config, items=items
             )
+            ignore = set(self._deselected + items)
+            removed_items = items_before.difference(ignore)
+            if removed_items:
+                # Explicitly deselect them.
+                # Calls the own hook again, but that is OK.
+                self.config.hook.pytest_deselected(items=removed_items)
```

Comment:
I guess with this change we can stop calling `pytest_deselected` explicitly in `cacheprovider`, `stepwise`, `mark`, etc?

Reply:
Yes.